pkg/provider/hub/hub.go

```diff
@@ -0,0 +1,209 @@
+package hub
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/traefik/traefik/v2/pkg/config/dynamic"
+	"github.com/traefik/traefik/v2/pkg/log"
+	"github.com/traefik/traefik/v2/pkg/provider"
+	"github.com/traefik/traefik/v2/pkg/safe"
+	ttls "github.com/traefik/traefik/v2/pkg/tls"
+)
+
+var _ provider.Provider = (*Provider)(nil)
+
+// DefaultEntryPointName the name of the default internal entry point.
+const DefaultEntryPointName = "traefik-hub"
+
+// Provider holds configurations of the provider.
+type Provider struct {
+	EntryPoint string `description:"Entrypoint that exposes data for Traefik Hub. It should be a dedicated one, and not used by any router." json:"entryPoint,omitempty" toml:"entryPoint,omitempty" yaml:"entryPoint,omitempty" export:"true"`
+	TLS        *TLS   `description:"TLS configuration for mTLS communication between Traefik and Hub Agent." json:"tls,omitempty" toml:"tls,omitempty" yaml:"tls,omitempty" export:"true"`
+
+	server *http.Server
+}
+
+// TLS holds TLS configuration to use mTLS communication between Traefik and Hub Agent.
+type TLS struct {
+	Insecure bool               `description:"Allows the Hub provider to run over an insecure connection for testing purposes." json:"insecure,omitempty" toml:"insecure,omitempty" yaml:"insecure,omitempty" export:"true"`
+	CA       ttls.FileOrContent `description:"Certificate authority to use for securing communication with the Agent." json:"ca,omitempty" toml:"ca,omitempty" yaml:"ca,omitempty" loggable:"false"`
+	Cert     ttls.FileOrContent `description:"Certificate to use for securing communication with the Agent." json:"cert,omitempty" toml:"cert,omitempty" yaml:"cert,omitempty" loggable:"false"`
+	Key      ttls.FileOrContent `description:"Key to use for securing communication with the Agent." json:"key,omitempty" toml:"key,omitempty" yaml:"key,omitempty" loggable:"false"`
+}
+
+// SetDefaults sets the default values.
+func (p *Provider) SetDefaults() {
+	p.EntryPoint = DefaultEntryPointName
+}
+
+// Init the provider.
+func (p *Provider) Init() error {
+	return nil
+}
+
+// Provide allows the hub provider to provide configurations to traefik using the given configuration channel.
+func (p *Provider) Provide(configurationChan chan<- dynamic.Message, _ *safe.Pool) error {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return fmt.Errorf("listener: %w", err)
+	}
+	port := listener.Addr().(*net.TCPAddr).Port
+
+	client, err := createAgentClient(p.TLS)
+	if err != nil {
+		return fmt.Errorf("creating Hub Agent HTTP client: %w", err)
+	}
+
+	p.server = &http.Server{Handler: newHandler(p.EntryPoint, port, configurationChan, p.TLS, client)}
+
+	go func() {
```

Comment:
```suggestion
	// TODO: this is going to be leaky (because no context to make it terminate)
	// if/when Provide lifecycle differs with Traefik lifecycle.
	go func() {
```